File: apps/web/src/features/booking/api/mappers.ts
/**
 * Data transformation functions for booking API responses.
 *
 * Mappers convert between:
 * - API/Supabase row format (snake_case)
 * - Domain model format (camelCase)
 *
 * This centralizes all data transformation logic for easier testing
 * and maintenance.
 */

import type { Booking } from '@housing-platform/types';

import type {
  BookingDetail,
  BookingDetailRow,
  BookingListItem,
  BookingListRow,
  BookingPriceSnapshotRow,
  BookingQuote,
  BookingQuoteRow,
} from '../model';
import { calculateServiceFeePercent } from '../model';

// ============================================================================
// Helper Functions
// ============================================================================

/**
 * Extract single item from Supabase relation (handles array vs object).
 */
export function extractFirstQuoteRow(
  data: BookingQuoteRow[] | null | undefined,
): BookingQuoteRow | null {
  return data?.[0] ?? null;
}

export function mapBookingRecord(data: unknown): Booking | null {
  if (typeof data !== 'object' || data === null || !('id' in data)) {
    return null;
  }

  return data as Booking;
}

export function getRelation<T>(value: T | T[] | null | undefined): T | null {
  if (value == null) {
    return null;
  }

  return Array.isArray(value) ? (value[0] ?? null) : value;
}

// ============================================================================
// Response → Domain Mappers
// ============================================================================

/**
 * Map quote response row to domain model.
 */
export function mapQuoteRow(row: BookingQuoteRow): BookingQuote {
  return {
    roomId: row.room_id,
    propertyId: row.property_id,
    bookingMode: row.booking_mode,
    nights: row.nights,
    rentKrw: row.rent_krw,
    serviceFeeKrw: row.service_fee_krw,
    totalKrw: row.total_krw,
    pricingVersion: row.pricing_version,
  };
}

/**
 * Map booking list row to domain model.
 * Returns null if required relations are missing.
 */
export function mapBookingListRow(row: BookingListRow): BookingListItem | null {
  const property = getRelation(row.properties);
  const room = getRelation(row.rooms);
  const snapshot = getRelation(row.booking_price_snapshots);

  if (!property || !room || !snapshot) {
    return null;
  }

  return {
    id: row.id,
    status: row.status as BookingListItem['status'],
    bookingType: row.booking_type as BookingListItem['bookingType'],
    checkIn: row.check_in,
    checkOut: row.check_out,
    guestCount: row.guest_count,
    propertyTitle: property.title,
    district: property.district,
    roomName: room.name,
    totalKrw: snapshot.total_krw,
    holdExpiresAt: row.hold_expires_at,
    createdAt: row.created_at,
  };
}

/**
 * Map booking detail row to domain model.
 * Returns null if required relations are missing.
 */
export function mapBookingDetailRow(row: BookingDetailRow): BookingDetail | null {
  const listItem = mapBookingListRow(row);
  const snapshot = getRelation(row.booking_price_snapshots) as BookingPriceSnapshotRow | null;

  if (!listItem || !snapshot) {
    return null;
  }

  return {
    ...listItem,
    customerNotes: row.customer_notes,
    rentKrw: snapshot.rent_krw,
    serviceFeeKrw: snapshot.service_fee_krw,
    serviceFeePercent: calculateServiceFeePercent(snapshot.rent_krw, snapshot.service_fee_krw),
    pricingVersion: snapshot.pricing_version,
    propertyId: row.property_id,
    roomId: row.room_id,
  };
}

// ============================================================================
// Domain → Request Mappers (if needed for mutations)
// ============================================================================

/**
 * Map booking form input to API request parameters.
 */
export function toCreateBookingRequest(input: {
  roomId: string;
  checkIn: string;
  checkOut: string;
  guestCount: number;
  customerNotes?: string;
}) {
  return {
    p_room_id: input.roomId,
    p_check_in: input.checkIn,
    p_check_out: input.checkOut,
    p_guest_count: input.guestCount,
    p_customer_notes: input.customerNotes ?? null,
  };
}

/**
 * Map quote input to API request parameters.
 */
export function toQuoteRequest(input: {
  roomId: string;
  checkIn: string;
  checkOut: string;
  guestCount: number;
}) {
  return {
    p_room_id: input.roomId,
    p_check_in: input.checkIn,
    p_check_out: input.checkOut,
    p_guest_count: input.guestCount,
  };
}

File: apps/backend-python/src/housing_platform/bookings/mappers.py
import uuid
from datetime import date, datetime

from housing_platform.bookings.schemas import (
    Booking,
    BookingDetail,
    BookingListItem,
    BookingMode,
    BookingQuote,
    BookingStatus,
    BookingType,
)
from housing_platform.db.models import Bookings


def calculate_service_fee_percent(rent_krw: int, service_fee_krw: int) -> float:
    if rent_krw <= 0:
        return 0.0
    return round((service_fee_krw / rent_krw) * 100)


def map_booking_quote(
    *,
    room_id: uuid.UUID,
    property_id: uuid.UUID,
    booking_mode: str,
    nights: int,
    rent_krw: int,
    service_fee_krw: int,
    total_krw: int,
    pricing_version: str,
) -> BookingQuote:
    return BookingQuote(
        room_id=room_id,
        property_id=property_id,
        booking_mode=BookingMode(booking_mode),
        nights=nights,
        rent_krw=rent_krw,
        service_fee_krw=service_fee_krw,
        total_krw=total_krw,
        pricing_version=pricing_version,
    )


def map_booking(booking: Bookings) -> Booking:
    return Booking(
        id=booking.id,
        customer_id=booking.customer_id,
        room_id=booking.room_id,
        property_id=booking.property_id,
        check_in=booking.check_in,
        check_out=booking.check_out,
        guest_count=booking.guest_count,
        status=BookingStatus(booking.status),
        booking_type=BookingType(booking.booking_type),
        hold_expires_at=booking.hold_expires_at,
        customer_notes=booking.customer_notes,
        approved_at=booking.approved_at,
        approved_by=booking.approved_by,
        cancelled_at=booking.cancelled_at,
        payment_retry_count=booking.payment_retry_count,
        created_at=booking.created_at,
        updated_at=booking.updated_at,
    )


def map_booking_list_item(
    *,
    booking_id: uuid.UUID,
    status: str,
    booking_type: str,
    check_in: date,
    check_out: date,
    guest_count: int,
    property_title: str,
    district: str,
    room_name: str,
    total_krw: int,
    hold_expires_at: datetime | None,
    created_at: datetime,
) -> BookingListItem:
    return BookingListItem(
        id=booking_id,
        status=BookingStatus(status),
        booking_type=BookingType(booking_type),
        check_in=check_in,
        check_out=check_out,
        guest_count=guest_count,
        property_title=property_title,
        district=district,
        room_name=room_name,
        total_krw=total_krw,
        hold_expires_at=hold_expires_at,
        created_at=created_at,
    )


def map_booking_detail(
    list_item: BookingListItem,
    *,
    customer_notes: str | None,
    rent_krw: int,
    service_fee_krw: int,
    pricing_version: str,
    property_id: uuid.UUID,
    room_id: uuid.UUID,
) -> BookingDetail:
    return BookingDetail(
        **list_item.model_dump(),
        customer_notes=customer_notes,
        rent_krw=rent_krw,
        service_fee_krw=service_fee_krw,
        service_fee_percent=calculate_service_fee_percent(rent_krw, service_fee_krw),
        pricing_version=pricing_version,
        property_id=property_id,
        room_id=room_id,
    )
